### PR TITLE
Azkaban CLI features

### PR DESCRIPTION
**Pull request by pranayhasan**
Create Project, Flow Status, Execute Flow and Cancel Flow

### CODE REVIEW

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanCreateProjectTask.groovy [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      logger.lifecycle("Enter Project description:");
+      String description = System.console().readLine(" Input>");
+      while (description.isEmpty()) {
+        logger.lifecycle("Enter a non-empty Project description");
+        description = System.console().readLine(" Input>");
+      }
+
+      URI createURI = new URIBuilder(azkProject.azkabanUrl)
+          .setPath("/manager")
+          .setParameter("session.id", sessionId)
```
Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanCreateProjectTask.groovy]: The values "session.id", "action", "name" etc are duplicated at various places. Can we separate them out to an Enum or a class and use that instead. Something like AzkabanParams.SESSION_ID or AzkabanParams.ACTION.

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanCreateProjectTask.groovy [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          .setParameter("action", "create")
+          .setParameter("name", azkProject.azkabanProjName)
+          .setParameter("description", description.trim().toString())
+          .build();
+
+      String response = AzkabanHelper.responseFromPOST(createURI);
+
+      if (response.toLowerCase().contains("error")) {
```
Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanCreateProjectTask.groovy]: Similar to above comment, can we have an AzkabanResponses enum or class for responses such as error etc?

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanExecuteFlowTask.groovy [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linkedin.gradle.azkaban;
+
+import org.apache.http.client.utils.URIBuilder;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException
```
Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanExecuteFlowTask.groovy]: Missing semicolons

File: hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanExecuteFlowTask.groovy [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+      List<String> flows = AzkabanHelper.fetchSortedFlows(new JSONObject(response), azkProject);
+
+      Set<String> indexSet;
+
+      if (AzkabanPlugin.interactive) {
+        logger.lifecycle("-----    ----------");
+        logger.lifecycle("INDEX    FLOWS LIST");
+        logger.lifecycle("-----    ----------");
+        for(int i=0; i<flows.size(); i++) {
```
Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanExecuteFlowTask.groovy]: flows.each?

Comment (nntnag17) [inline hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanExecuteFlowTask.groovy]: or if you need index eachWithIndex?

**Comment by pranayhasan** [comment]
Want to know if there are any quick additions/IO changes. Would cleanup once the view is finalized.

**Comment by convexquad** [comment]
@pranayhasan Since this is a larger RB, give me a few more days to work on the review

**Comment by convexquad** [comment]
@pranayhasan @nntnag17 I want to ask that you make a major design change with this PR. I don't want to repeat the same mistake as with the azkaban-cli (internal to LinkedIn) multiproduct, where it is hard to re-use the code because it assumes that you are interacting with it in a specific way.

I would like to ask you to create a new Gradle subproject for interacting with Azkaban that just contains a series of library methods with high-level API's, e.g. createAzkabanProject(AzkabanProject azkabanProject, String sessionId) that don't assume anything about Gradle or the Hadoop DSL / Plugin.

Then leave all the Gradle tasks in place in the hadoop-plugin project. Probably there will still be the same classes that there are now (AzkabanCreateProjectTask, etc.), however instead of having the code in them that creates the project, they should call the API from the other subproject.

**Comment by convexquad** [comment]
Second comment - can you include as many unit tests as you can for any standalone helper functions (i.e. the functions that don't make remote calls to Azkaban). For the functions that actually call Azkaban, testing will be more difficult. We should develop tests for these methods, but I think we can do it as a separate PR.

**Comment by pranayhasan** [comment]
Thanks @convexquad for the recommendations. I agree with the concerns raised by you, as I haven't been able to reuse the Azkaban-CLI though it is inherently calling the same API. Would work on refactoring to make the API calls independent to that of hadoop-plugin. I will continue to commit on the existing PR and ping you once I'm done.

**Comment by convexquad** [comment]
@pranayhasan Do you have an ETA on the refactoring work for this Pull Request? It would be an awesome enhancement for the Hadoop Plugin

**Comment by pranayhasan** [comment]
@convexquad I had other prioritized work, so had to put this on backlog. I started refactoring it. You can expect updated PR by early next week.

**Comment by pranayhasan** [comment]
Opened a new PR discarding this.

**Pull request closed without merging**